lib/archethic/p2p.ex

```diff
@@ -732,162 +733,114 @@ defmodule Archethic.P2P do
-          if previous_result do
-            do_quorum([previous_result | results], conflict_resolver)
-          else
-            List.first(results)
-          end
-
-        with true <- acceptance_resolver.(quorum_result),
-             nil <- previous_result do
-          do_quorum_read(
-            rest,
-            message,
-            conflict_resolver,
-            acceptance_resolver,
-            timeout,
-            consistency_level,
-            acceptance_failures_count,
-            quorum_result
-          )
-        else
-          false ->
-            do_quorum_read(
-              rest,
-              message,
-              conflict_resolver,
-              acceptance_resolver,
-              timeout,
-              consistency_level,
-              acceptance_failures_count + 1,
-              previous_result
-            )
+        true ->
+          {:error, :acceptance_failed}
+      end
+    end)
+  end
 
-          _ ->
-            {:ok, quorum_result}
-        end
+  defp send_message_and_filter_results(nodes, message, timeout) do
+    Task.Supervisor.async_stream_nolink(
+      TaskSupervisor,
+      nodes,
+      &{&1.first_public_key, send_message(&1, message, timeout)},
+      ordered: false,
+      on_timeout: :kill_task,
+      timeout: timeout + 2000
+    )
+    |> Stream.filter(&match?({:ok, {_, {:ok, _}}}, &1))
+    |> Stream.map(fn {:ok, {node_public_key, {:ok, res}}} -> {node_public_key, res} end)
+    |> Enum.to_list()
+  end
+
+  # returns ok if we have 2 results
+  # we should probably use the consistency level
+  defp enough_results?(nil, results_by_node) do
+    case results_by_node do
+      [] -> {:error, nil}
+      [{_node, result}] -> {:error, result}
+      _ -> :ok
+    end
+  end
 
-      _ ->
-        results = if previous_result != nil, do: [previous_result | results], else: results
-        quorum_result = do_quorum(results, conflict_resolver)
-
-        if acceptance_resolver.(quorum_result) do
-          {:ok, quorum_result}
-        else
-          do_quorum_read(
-            rest,
-            message,
-            conflict_resolver,
-            acceptance_resolver,
-            timeout,
-            consistency_level,
-            acceptance_failures_count + 1,
-            previous_result
-          )
-        end
+  defp enough_results?(previous_result, results_by_node) do
+    case results_by_node do
+      [] -> {:error, previous_result}
+      _ -> :ok
     end
   end
 
-  defp do_quorum(results, conflict_resolver) do
-    distinct_elems = Enum.dedup(results)
+  defp resolve_conflicts(nil, results_by_node, conflict_resolver) do
+    results_by_node
+    |> Enum.map(&elem(&1, 1))
+    |> Enum.dedup()
+    |> conflict_resolver.()
+  end
+
+  defp resolve_conflicts(previous_result, results_by_node, conflict_resolver) do
+    results_by_node
+    |> Enum.map(&elem(&1, 1))
+    |> Kernel.++([previous_result])
+    |> Enum.dedup()
+    |> conflict_resolver.()
```

Comment:
Suggestion:
Here you could avoid running the conflict resolver if there is only in distinct result after the dedup (as it is in the current implementation)